htsinfer/infer_adapter.py

```diff
@@ -0,0 +1,203 @@
+    ('CCGAGCCCACGAGAC', 8),
+    ('CCGGTTCTTCCCTGC', 9),
+    ('CGACAGGTTCAGAGT', 10),
+    ('CGGTCTCGGCATTCC', 11),
+    ('CTGTAGGCACCATCA', 12),
+    ('GAAUUCCACCACGUU', 13),
+    ('GATCGGAAGAGCACA', 14),
+    ('GATCGGAAGAGCGGT', 15),
+    ('GATCGGAAGAGCTCG', 16),
+    ('GATCGTCGGACTGTA', 17),
+    ('GCCTTGGCACCCGAG', 18),
+    ('GTCTCGTGGGCTCGG', 19),
+    ('GTGACTGGAGTTCAG', 20),
+    ('GUUCAGAGUUCUACA', 21),
+    ('TCGGACTGTAGAACT', 22),
+    ('TCGTATGCCGTCTTC', 23),
+    ('TCGTCGGCAGCGTCA', 24),
+    ('TGGAATTCTCGGGTG', 25),
+    ('UCGUAUGCCGUCUUC', 26),
+    ('AGATCGGAAGAGCGT', 27)
+]
+
+
+def make_aho_auto(
+    adapters: List[Tuple[str, int]]
+):
+    """Adding all adapters names into trie"""
+    LOGGER.debug("Creating trie")
+    trie = ahc.Automaton()
+    for (adapter, tag) in adapters:
+        trie.add_word(adapter, (tag, adapter))
+
+    trie.make_automaton()
+    return trie
+
+
+TRIE = make_aho_auto(ADAPTERS)
+
+
+def infer(
+    file_1: str,
+    file_2: str = None,
+    max_records: int = 100000,
+    min_match: float = 10,
+    factor: float = 2
+) -> Tuple[str, str]:
+    """Infers adapter information for one or two fastq files.
+
+    Args:
+        file_1 (str) : File path to read/first mate library.
+        file_2 (str) : File path to second mate library.
+        max_records: Limit processing to the indicated number of records,
+            starting from the first record.
+        min_match (float) : minimum match percentage that first adapter needs
+        to have.
+        factor (float) : factor by which first adapter is greater than the
+        second adapter.
+
+    Returns:
+        Type of Adapter that is present in library.
+    """
+    # Process file 1
+    LOGGER.debug(f"Processing file 1: {file_1}")
+    result_1 = process_fastq_file(file_1, max_records, min_match, factor)
+
+    # Process file 2
+    result_2 = "not_available"
+    if file_2:
+        LOGGER.debug(f"Processing file 2: {file_2}")
+        result_2 = process_fastq_file(file_2, max_records, min_match, factor)
+
+    LOGGER.debug("Returning results...")
+    return (result_1, result_2)
+
+
+def process_fastq_file(
+    file: str,
+    max_records: int = 10000,
+    min_match: float = 10,
+    factor: float = 2
+) -> str:
+    """Process adapters count info.
+
+    Args:
+        file (str) : File path to read/first mate library.
+        max_records: Limit processing to the indicated number of records,
+            starting from the first record.
+        min_match (float) : minimum match percentage that first adapter needs
+        to have.
+        factor (float) : factor by which first adapter is greater than the
+        second adapter.
+
+    Returns:
+        Adapter information
+    """
+    _open = partial(
+        gzip.open, mode='rt'
+    ) if file.endswith(".gz") else open
+
+    try:
+        LOGGER.debug("Opening file...")
+        with _open(file) as _file:  # type: ignore
+
+            adapter_counts: Dict[str, float] = {}
```

Comment:
Why not define all of these before openeing the file object? Just put the code in the `with` context manager that you really need the file open for